src/agent_observer.py

```diff
@@ -0,0 +1,91 @@
+# Copyright 2024 Canonical Ltd.
+# See LICENSE file for licensing details.
+
+"""The agent relation observer module."""
+
+
+import logging
+
+import ops
+
+import service
+from charm_state import AGENT_RELATION, State, get_agent_interface_dict_from_metadata
+
+logger = logging.getLogger()
+
+
+class Observer(ops.Object):
+    """The Jenkins agent relation observer."""
+
+    def __init__(
+        self,
+        charm: ops.CharmBase,
+        state: State,
+        jenkins_agent_service: service.JenkinsAgentService,
+    ):
+        """Initialize the observer and register event handlers.
+
+        Args:
+            charm: The parent charm to attach the observer to.
+            state: The charm state.
+            jenkins_agent_service: Service manager that controls Jenkins agent service.
+        """
+        super().__init__(charm, "agent-observer")
+        self.charm = charm
+        self.state = state
+        self.jenkins_agent_service = jenkins_agent_service
+        charm.framework.observe(
+            charm.on[AGENT_RELATION].relation_joined, self._on_agent_relation_joined
+        )
+        charm.framework.observe(
+            charm.on[AGENT_RELATION].relation_changed, self._on_agent_relation_changed
+        )
+        charm.framework.observe(
+            charm.on[AGENT_RELATION].relation_departed, self._on_agent_relation_departed
+        )
+
+    def _on_agent_relation_joined(self, event: ops.RelationJoinedEvent) -> None:
+        """Handle agent relation joined event.
+
+        Args:
+            event: The event fired when an agent has joined the relation.
+        """
+        self.charm.unit.status = ops.MaintenanceStatus(
+            f"Setting up '{event.relation.name}' relation."
+        )
+
+        relation_data = get_agent_interface_dict_from_metadata(self.state.agent_meta)
+        logger.debug("Setting agent relation unit data: %s", relation_data)
+        event.relation.data[self.charm.unit].update(relation_data)
+
+    def _on_agent_relation_changed(self, _: ops.RelationChangedEvent) -> None:
+        """Handle agent relation changed event.
+
+        Raises:
+            RuntimeError: when the service fails to properly start.
+        """
+        # Check if the jenkins agent service has started and set agent ready.
+        if self.jenkins_agent_service.is_active:
+            logger.warning("Given agent already registered. Skipping.")
```

Comment:
Does it really make sense to warn? It doesn't look like this would be abnormal imho

Reply:
I changed this to `logger.error`